README.md
# interactive-brokers-flex-rs
Tools to assist with IB Flex reports and Ledger-cli comparison

This crate contains a CLI application and is also a library which assists working with Flex Reports from Interactive Brokers. It simplifies the Flex Query download and compares the downloaded transactions (distributions and tax) to the records in Ledger, identifying missing ones.

There are several components in the package:

- The command-line application (CLI) `ibflex`, provides all the features of the `ibflex` library
- `ibflex` library exposes the following functionality:
  - downloads the IB Flex Query report
  - parses IB Flex Query report
  - runs Ledger-cli to retrieve the transactions in the last 60 days
  - compares the Cash Transactions from the downloaded Flex Query .xml report to the Ledger transactions
- `as-symbols` [crate](https://crates.io/crates/as-symbols) provides the Symbol mapping between IB Flex report and Ledger. I.e. symbol `VHYL` in the report is `VHYL_AS` in Ledger.

The project started as a rewrite of my Python scripts and is intended to be expanded as needed, to parse Flex Queries.

# Configuration

To view the current configuration, run

```
ibflex cfg
```

The config file will be created automatically if it does not exist.

To edit the values, use any text editor.

## Symbols Configuration

See `as-symbols` [crate](https://crates.io/crates/as-symbols) for instructions on how to set up the symbols data file.
At the moment this is required for the symbol mapping. The symbols in IB (i.e. VHYL) may be mapped to a different symbol in Ledger (i.e. VHYL_AS).

# Usage

Downloading the Flex Query report requires Query Id and the Token. These can be passed in several ways:

1) as parameters to the download (Dl) method. Use parameters `--queryid` and `--token`.
2) as environment variables:

  The application will read the values from environment variables:

  - `IBFLEX_TOKEN`
  - `IBFLEX_QUERYID`

3) in the configuration file: The application will read the configuration file `ibflex.toml`, which is located in the current directory. See the section below.

## Download

The required parameters for downloading the Flex report are:

- `flex_query_id`
- `ib_token`

Once this is set up, invoke the CLI:

```
ifblex dl
```

This will save the report in the current directory. The filename will contain today's date.

## Comparison

To compare the transactions, run

```
ibflex cmp
```

This will compare the downloaded IB transactions to the transactions in Ledger. The new Dividend and Tax transactions will be reported as New. The other transactions will be reported as Skipped.

# Changelog

See [Changelog](CHANGELOG.md)

# Credits

- [ibflex](https://github.com/csingley/ibflex) Python library

# License

See [LICENSE](LICENSE) file.


CHANGELOG.md
# Changelog

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.3.0] - 2023-09-22

### Changed

- using the exchange/namespace to uniquely identify the symbol for comparison. Used for the symbols that are the same but on different exchanges.

### Security

- updated dependencies

## [1.2.1] - 2023-07-01

### Added

- showing more details in case of an invalid amount

### Security

- updated dependencies

## [1.2.0] - 2023-03-16

### Added

- parsing the payments in-lieu of dividends

### Removed

- the command to display the configuration
- the configuration file (ibflex.toml) functionality. The parameters should always be passed directly to a command.

### Security

- updated dependencies

## [1.1.0] - 2023-03-10

### Added

- an option to specify the ledger file path

### Changed

- `date-format` and `wide` display are now set automatically

### Removed

- the option to specify the ledger init file

### Security

- updated dependencies

## [0.11.0] - 2023-02-21

### Changed

- changed Config command to Setup, to free the "c" shortcut for Compare

### Security

- updated dependencies

## [0.10.0] - 2023-02-07

### Added

- aligning amounts in the comparison display
- the comparison transactions now also sorted by type

### Changed

- flex query id and token are now optional configuration settings

## [0.9.2] - 2023-02-06

### Changed

- comparison sort uses report date / date / symbol for ordering.

## [0.9.1] - 2023-02-06

### Added

- a parameter to specify a config file for the Compare command

### Changed

- sorting IB transactions by the report date

### Security

- updated dependencies

### Changed

- using the earliest date from the Flex report to time-box Ledger transactions

### Removed

- `comparison_date` parameter, which is now redundant

## [0.7.1] - 2023-01-25

### Added

- added `comparison_date` argument to run comparison on arbitrary dates.

### Fixed

- using IB transaction book date when comparing using `--effective`.

## [0.7.0] - 2023-01-25

### Added

- `--effective` argument, to use Ledger's effective dates

### Changed

- returning the comparison result string from `compare`

### Security

- updated dependencies

## [0.6.2] - 2023-01-20

### Changed

- using (ib_symbol, ledger_symbol) hashmap for comparison

## [0.6.1] - 2023-01-20

### Added

- displays the parsing errors, showing the exact location of the erroneous record

## [0.6.0] - 2023-01-20

### Added

- the path to the symbols file parameter for `compare`

### Changed

- using `as-symbols` package instead of `pricedb`. The symbols are stored in a CSV file.

### Security

- updated dependencies

## [0.5.1] - 2023-01-17

### Changed

- The member names in FlexQueryResponse are now Rust's standard snake_case.

## [0.5] - 2023-01-17

### Changed

- The member names in FlexStatementResponse are now Rust's standard snake_case.

### Security

- updated dependencies

## [0.4.4] - 2023-01-11

### Fixed

- skipping transaction other than dividends and withholding tax

### Changed

- exiting comparison if no new IB transactions found

## [0.4.3] - 2023-01-11

### Fixed

- Skipping transaction other than dividends and withholding tax.

### Security

- updated dependencies

## [0.4.2] - 2023-01-05

### Added

- Expanded instructions in the ReadMe file. Added the section for the comparison command.

### Fixed

- Ignoring all but distributions and tax. Using the new enum for type comparison.

### Security

- Updated tokio dependency.

## [0.4.1] - 2023-01-05

### Added

- Display the path to the IB Flex report .xml file used for comparison.

### Fixed

- Comparing Ledger tx date to Report Date in Flex Query.

## [0.4.0] - 2023-01-04

The first acceptable version.

### Changed

- Comparison tests pass


src/compare.rs
/*!
 * Compares the downloaded Flex report with Ledger
 */

use std::{
    collections::HashMap,
    ops::Mul,
    path::{Path, PathBuf},
};

use anyhow::{Error, Ok};
use as_symbols::SymbolMetadata;
use rust_decimal::Decimal;

use crate::{
    flex_enums::{cash_action, CashAction},
    flex_query::{CashTransaction, FlexQueryResponse},
    flex_reader::load_report,
    ledger_runner::{self, get_ledger_start_date},
    model::CommonTransaction,
    ISO_DATE_FORMAT,
};

pub const TRANSACTION_DAYS: u8 = 60;

/**
 * Compares transactions in the downloaded IB Flex report to Ledger.
 */
pub fn compare(params: CompareParams) -> anyhow::Result<String> {
    log::debug!("comparing distributions, params: {:?}", params);
    // let cfg = get_cmp_config(&params);

    // get_ib_report_tx
    let mut ib_txs = get_ib_tx(&params);
    log::debug!("Found {} IB transactions", ib_txs.len());
    if ib_txs.len() == 0 {
        let msg = "No new IB transactions found. Exiting...\n";
        print!("{}", msg);
        return Ok(msg.into());
    }

    // sort IB records by dates, symbol, type
    ib_txs.sort_unstable_by_key(|tx| {
        (
            tx.report_date.to_owned(),
            tx.date,
            tx.symbol.to_owned(),
            tx.r#type.to_owned(),
        )
    });
    log::debug!("sorted: {:?}", ib_txs);

    // identify the start date for the tx range:
    let start_date = get_oldest_ib_date(&ib_txs, params.effective_dates);

    // get_ledger_tx
    let ledger_txs = ledger_runner::get_ledger_tx(
        params.ledger_journal_file,
        start_date,
        params.effective_dates,
    );
    log::debug!("Found {} Ledger transactions", ledger_txs.len());

    // compare
    let result = compare_txs(ib_txs, ledger_txs, params.effective_dates)?;

    Ok(result)
}

fn compare_txs(
    ib_txs: Vec<CommonTransaction>,
    ledger_txs: Vec<CommonTransaction>,
    use_effective_date: bool,
) -> anyhow::Result<String> {
    let mut result = String::default();

    for ibtx in ib_txs {
        log::debug!(
            "Searching for matches for ib tx: {:?}\n among {:?}",
            ibtx,
            ledger_txs
        );

        let ib_comparison_date = get_comparison_date(&ibtx, use_effective_date);
        log::debug!("using ib date for comparison: {:?}", ib_comparison_date);

        let matches: Vec<&CommonTransaction> = ledger_txs
            .iter()
            .filter(|tx| {
                // Compare:
                tx.date.date().format(ISO_DATE_FORMAT).to_string() == ib_comparison_date
                    && tx.symbol == ibtx.symbol
                    && tx.amount == ibtx.amount.mul(Decimal::NEGATIVE_ONE)
                    && tx.currency == ibtx.currency
                    && tx.r#type == ibtx.r#type
            })
            .collect();
        
        // log::debug!("date: {:?}", ledger_txs[0].date.date().format(ISO_DATE_FORMAT).to_string());
        log::debug!("matching ledger txs: {:?}", matches);

        if matches.is_empty() {
            let output = format!("New: {}\n", ibtx);
            print!("{}", output);
            result.push_str(&output);
        }
    }
    println!("Complete.");

    Ok(result)
}

fn get_comparison_date(ibtx: &CommonTransaction, use_effective_date: bool) -> String {
    match use_effective_date {
        true => ibtx.date.format(ISO_DATE_FORMAT).to_string(),
        false => ibtx.report_date.to_owned(), // actual date
    }
}

/// Finds the date of the oldest transaction in the report.
/// This date is to be used for time-boxing Ledger report.
fn get_oldest_ib_date(ib_txs: &Vec<CommonTransaction>, use_effective_date: bool) -> String {
    if ib_txs.is_empty() {
        return get_ledger_start_date(None);
    }

    //ib_txs.sort_unstable_by_key(|ibtx| get_comparison_date(&ibtx, use_effective_date) );
    let oldest_date_record = ib_txs
        .iter()
        .min_by_key(|ibtx| get_comparison_date(&ibtx, use_effective_date))
        .expect("got oldest date");

    log::debug!("oldest tx: {:?}", oldest_date_record);

    get_comparison_date(&oldest_date_record, use_effective_date)
}

/// Load the symbol mappings.
/// The resulting hashmap is <symbol, ledger symbol>.
fn load_symbols(path: &PathBuf) -> Result<HashMap<String, String>, Error> {
    log::debug!("loading symbols from {:?}", path);

    // confirm the path exists
    let real_path = Path::new(path);
    if !real_path.exists() {
        panic!("The symbols file {:?} does not exist!", path);
    }

    let securities = as_symbols::read_symbols(path)
        .expect("Parsed symbols")
        .iter()
        .map(|sym| map_symbols(sym))
        .collect();
    Ok(securities)
}

/// Maps the SymbolMetadata into a hashmap of (ib_symbol, ledger_symbol) records.
fn map_symbols(meta: &SymbolMetadata) -> (String, String) {
    (
        match &meta.ib_symbol {
            Some(ib_sym) => ib_sym.to_owned(),
            None => {
                let Some(ref namespace) = meta.namespace
                    else { panic!("Missing namespace found!") };

                format!("{}:{}", namespace, meta.symbol)
            },
        },
        match &meta.ledger_symbol {
            Some(ldg_sym) => ldg_sym.to_owned(),
            None => meta.symbol.to_owned(),
        },
    )
}

/**
Returns transactions from the Flex Report, for comparison.
symbols is a HashMap of symbol rewrites.
*/
fn get_ib_tx(cfg: &CompareParams) -> Vec<CommonTransaction> {
    let ib_txs = read_flex_report(cfg);

    convert_ib_txs_into_common(ib_txs, cfg.symbols_path.as_str())
}

/// Converts IB CashTransaction XML record into a Common Transaction.
fn convert_ib_txs_into_common(ib_txs: Vec<CashTransaction>, symbols_path_str: &str) -> Vec<CommonTransaction> {
    // load symbols. Need a mapping to the ledger symbols for comparison.
    let symbols_path = PathBuf::from(symbols_path_str);
    let symbols = load_symbols(&symbols_path).unwrap();
    log::debug!("symbols loaded: {:?}", symbols);

    let mut txs: Vec<CommonTransaction> = vec![];

    let to_include = [
        CashAction::WhTax.to_string(),
        CashAction::Dividend.to_string(),
        CashAction::PaymentInLieu.to_string()
    ];
    log::debug!("to include: {:?}", to_include);

    for tx in ib_txs {
        log::debug!(
            "Converting ib tx: {:?} {:?} ({:?})",
            tx.symbol,
            tx.r#type,
            cash_action(&tx.r#type)
        );

        // skip any not matching the expected types.
        if !to_include.contains(&cash_action(&tx.r#type)) {
            log::debug!("Skipping. Wrong type ({:?})", &tx.r#type);
            println!("Skipped: {}", tx);
            continue;
        }

        let mut ltx: CommonTransaction = (&tx).into();

        // use adjusted symbols
        if symbols.contains_key(&ltx.symbol) {
            // log::debug!("adjusted symbol: {} -> {}", ltx.symbol, symbols[&ltx.symbol]);
            ltx.symbol = symbols[&ltx.symbol].to_owned();
        }

        txs.push(ltx);
    }

    txs
}

/**
 * Reads the Cash Transaction records from the Flex Report.
 * Sorts by date/time, symbol, type.
 */
fn read_flex_report(cfg: &CompareParams) -> Vec<CashTransaction> {
    let content = load_report(cfg.flex_report_path.to_owned(), cfg.flex_reports_dir.to_owned());
    let response = FlexQueryResponse::from(content);

    let mut ib_txs = response
        .flex_statements
        .flex_statement
        .cash_transactions
        .cash_transaction;

    // txs.sort(key=operator.attrgetter("dateTime", "symbol", "type.name"))
    ib_txs.sort_unstable_by_key(|ct| {
        (
            ct.date_time.to_owned(),
            ct.symbol.to_owned(),
            ct.r#type.to_owned(),
        )
    });

    ib_txs
}

/**
 * Parameters for comparing the IB Flex report and Ledger report.
 */
#[derive(Debug)]
pub struct CompareParams {
    pub flex_report_path: Option<String>,
    pub flex_reports_dir: Option<String>,
    pub ledger_journal_file: Option<String>,
    pub symbols_path: String,
    pub effective_dates: bool,
}

// Tests

#[cfg(test)]
mod tests {
    use std::path::PathBuf;

    use super::{compare, load_symbols};
    use crate::{
        compare::{convert_ib_txs_into_common, CompareParams},
        flex_query::CashTransaction,
        test_fixtures::*,
    };

    /// Load symbols through PriceDb.
    #[test]
    fn symbols_load_test() {
        let symbol_path = PathBuf::from("tests/symbols.csv");

        let symbols = load_symbols(&symbol_path).expect("symbols loaded");

        assert!(!symbols.is_empty());
    }

    #[rstest::rstest]
    #[test_log::test]
    fn test_convert_ib_txs(cash_transactions: Vec<CashTransaction>) {
        let symbols_path = "tests/symbols.csv";

        let ib_tx = convert_ib_txs_into_common(cash_transactions, symbols_path);

        assert!(!ib_tx.is_empty());
    }

    #[rstest::rstest]
    #[test_log::test]
    fn test_compare(cmp_params: CompareParams) {
        println!("comparing using: {:?}", cmp_params);

        let actual = compare(cmp_params);

        assert!(!actual.is_err());
    }

    /// tax adjustments come on one day and match several records in the past year.
    /// The report date needs to be matched to the effective date in this case,
    /// in addition to the transaction date/transaction date.
    #[test_log::test]
    fn test_compare_w_multiple_matches() {
        let cmp_params = CompareParams {
            flex_report_path: Some("tests/tax_adj_report.xml".into()),
            flex_reports_dir: None,
            ledger_journal_file: Some("tests/tax_adj_journal.ledger".into()),
            symbols_path: "tests/symbols.csv".to_owned(),
            effective_dates: false,
        };
        let actual = compare(cmp_params).unwrap();

        println!("result: {:?}", actual);

        let expected = "";

        assert_eq!(expected, actual);
    }

    /// Same test but using effective dates.
    #[test_log::test]
    fn test_compare_w_multiple_matches_effective_dates() {
        let cmp_params = CompareParams {
            flex_report_path: Some("tests/tax_adj_report.xml".into()),
            flex_reports_dir: None,
            ledger_journal_file: Some("tests/tax_adj_journal.ledger".into()),
            symbols_path: "tests/symbols.csv".into(),
            effective_dates: true,
        };
        let actual = compare(cmp_params).unwrap();

        println!("result: {:?}", actual);

        let expected = r#"New: 2023-01-24/2022-04-01 BBN     WhTax       0.66 USD, BBN(US09248X1000) CASH DIVIDEND USD 0.1229 PER SHARE - US TAX
New: 2023-01-24/2022-04-01 BBN     WhTax      -0.53 USD, BBN(US09248X1000) CASH DIVIDEND USD 0.1229 PER SHARE - US TAX
New: 2023-01-24/2022-04-30 BBN     WhTax       0.66 USD, BBN(US09248X1000) CASH DIVIDEND USD 0.1229 PER SHARE - US TAX
New: 2023-01-24/2022-04-30 BBN     WhTax      -0.53 USD, BBN(US09248X1000) CASH DIVIDEND USD 0.1229 PER SHARE - US TAX
"#;

        /* These are in the ledger file:
        New: 2023-01-24/2022-03-01 BBN     WhTax    0.66 USD, BBN(US09248X1000) CASH DIVIDEND USD 0.1229 PER SHARE - US TAX
        New: 2023-01-24/2022-03-01 BBN     WhTax   -0.53 USD, BBN(US09248X1000) CASH DIVIDEND USD 0.1229 PER SHARE - US TAX
         */

        assert_eq!(expected, actual);
    }

    #[test_log::test]
    fn test_tcf() {
        let cmp_params = CompareParams {
            flex_report_path: Some("tests/tcf.xml".into()),
            flex_reports_dir: None,
            ledger_journal_file: Some("tests/tcf.ledger".into()),
            symbols_path: "tests/symbols.csv".into(),
            effective_dates: false,
        };
        let actual = compare(cmp_params).unwrap();

        let expected = "";

        assert_eq!(expected, actual);
    }

    #[test]
    fn test_in_lieu() {
        let params = CompareParams {
            flex_report_path: Some("tests/in-lieu.xml".to_owned()),
            flex_reports_dir: None,
            ledger_journal_file: Some("tests/in-lieu.ledger".into()),
            symbols_path: "tests/symbols.csv".into(),
            effective_dates: false,
        };

        let actual = compare(params).unwrap();

        //let expected = r#"New: 2023-03-15/2023-03-16 EXXW    PaymentInLieu    3.74 EUR, EXXW(DE000A0H0744) PAYMENT IN LIEU OF DIVIDEND (Mixed Income)\n"#;
        let expected = "";
        
        assert_eq!(expected, actual);
    }

    #[test_log::test]
    fn test_same_symbols_different_exchange() {
        let cmp_params = CompareParams {
            flex_report_path: Some("tests/same_symbol.xml".into()),
            flex_reports_dir: None,
            ledger_journal_file: Some("tests/same_symbol.ledger".into()),
            symbols_path: "tests/symbols.csv".into(),
            effective_dates: false,
        };
        let actual = compare(cmp_params).unwrap();

//         let expected = r#"New: 2023-09-14/2023-09-15 ARCA:SDIV Dividend    5.04 USD, SDIV(US37960A6698) CASH DIVIDEND USD 0.21 PER SHARE (Ordinary Dividend)
// New: 2023-09-21/2023-09-22 BVME.ETF:SDIV    Dividend   10.26 USD, SDIV(IE00077FRP95) CASH DIVIDEND USD 0.09 PER SHARE (Mixed Income)"#;
        let expected = "";

        assert_eq!(expected, actual);
    }
}
